File: docs/beginner/module-06-filters-sorting/code-samples/python/07_complex_filters.py
#!/usr/bin/env python3
"""
Azure AI Search - Complex Filters

This script demonstrates advanced filtering operations using Azure AI Search,
including collection filtering, nested conditions, and complex logical combinations.

Key Features:
- Collection filtering with any() and all() functions
- Nested condition optimization
- Advanced logical combinations
- Filter logic tree visualization
- Performance optimization for complex filters
- Real-world complex filtering scenarios

Prerequisites:
- Azure AI Search service configured
- Sample data with collection fields loaded
- Environment variables set in .env file
"""

import os
import time
import re
from typing import List, Dict, Any, Optional, Union
from dataclasses import dataclass
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

@dataclass
class FilterCondition:
    """Represents a single filter condition."""
    field: str
    operator: str
    value: Any
    collection_function: Optional[str] = None  # 'any' or 'all'

@dataclass
class FilterGroup:
    """Represents a group of filter conditions with logical operator."""
    conditions: List[Union[FilterCondition, 'FilterGroup']]
    logical_operator: str = 'and'  # 'and' or 'or'

class ComplexFilterBuilder:
    """Builds and manages complex filter expressions for Azure AI Search."""
    
    def __init__(self):
        """Initialize the complex filter builder."""
        self.search_client = self._initialize_client()
        self.filter_templates = self._load_filter_templates()
    
    def _initialize_client(self) -> SearchClient:
        """Initialize Azure AI Search client."""
        try:
            endpoint = os.getenv('SEARCH_ENDPOINT')
            api_key = os.getenv('SEARCH_API_KEY')
            index_name = os.getenv('INDEX_NAME')
            
            if not all([endpoint, api_key, index_name]):
                raise ValueError("Missing required environment variables")
            
            credential = AzureKeyCredential(api_key)
            client = SearchClient(endpoint, index_name, credential)
            
            print(f"✅ Connected to Azure AI Search")
            print(f"📍 Endpoint: {endpoint}")
            print(f"📊 Index: {index_name}")
            
            return client
            
        except Exception as e:
            print(f"❌ Configuration error: {e}")
            raise
    
    def _load_filter_templates(self) -> Dict[str, str]:
        """Load common complex filter templates."""
        return {
            'collection_any': "{collection}/any(item: item eq '{value}')",
            'collection_all': "{collection}/all(item: item eq '{value}')",
            'collection_contains': "{collection}/any(item: contains(item, '{value}'))",
            'nested_and': "({condition1}) and ({condition2})",
            'nested_or': "({condition1}) or ({condition2})",
            'range_with_collection': "({range_condition}) and {collection}/any(item: item eq '{value}')",
            'multi_collection': "{collection1}/any(item: item eq '{value1}') and {collection2}/any(item: item eq '{value2}')"
        }
    
    def build_collection_filter(self, collection_field: str, values: List[str], 
                              function: str = 'any', operator: str = 'eq') -> str:
        """
        Build a collection filter expression.
        
        Args:
            collection_field: Name of the collection field
            values: List of values to match
            function: 'any' or 'all'
            operator: Comparison operator ('eq', 'contains', etc.)
            
        Returns:
            OData filter expression
        """
        if not values:
            raise ValueError("At least one value must be provided")
        
        if function not in ['any', 'all']:
            raise ValueError("Function must be 'any' or 'all'")
        
        conditions = []
        for value in values:
            if operator == 'eq':
                condition = f"item eq '{value}'"
            elif operator == 'contains':
                condition = f"contains(item, '{value}')"
            elif operator == 'startswith':
                condition = f"startswith(item, '{value}')"
            elif operator == 'endswith':
                condition = f"endswith(item, '{value}')"
            else:
                raise ValueError(f"Unsupported operator: {operator}")
            
            conditions.append(condition)
        
        # Combine conditions with OR for multiple values
        combined_condition = " or ".join(conditions)
        
        return f"{collection_field}/{function}(item: {combined_condition})"
    
    def build_nested_filter(self, conditions: List[FilterCondition], 
                          logical_operator: str = 'and') -> str:
        """
        Build a nested filter with multiple conditions.
        
        Args:
            conditions: List of filter conditions
            logical_operator: 'and' or 'or'
            
        Returns:
            OData filter expression
        """
        if not conditions:
            raise ValueError("At least one condition must be provided")
        
        filter_parts = []
        
        for condition in conditions:
            if condition.collection_function:
                # Collection filter
                if condition.operator == 'eq':
                    item_condition = f"item eq '{condition.value}'"
                elif condition.operator == 'contains':
                    item_condition = f"contains(item, '{condition.value}')"
                else:
                    item_condition = f"item {condition.operator} '{condition.value}'"
                
                filter_part = f"{condition.field}/{condition.collection_function}(item: {item_condition})"
            else:
                # Regular filter
                if isinstance(condition.value, str):
                    filter_part = f"{condition.field} {condition.operator} '{condition.value}'"
                else:
                    filter_part = f"{condition.field} {condition.operator} {condition.value}"
            
            filter_parts.append(filter_part)
        
        return f" {logical_operator} ".join(filter_parts)
    
    def build_complex_filter_tree(self, filter_group: FilterGroup) -> str:
        """
        Build a complex filter from a hierarchical filter tree.
        
        Args:
            filter_group: Root filter group
            
        Returns:
            OData filter expression
        """
        parts = []
        
        for item in filter_group.conditions:
            if isinstance(item, FilterCondition):
                # Single condition
                if item.collection_function:
                    if item.operator == 'eq':
                        item_condition = f"item eq '{item.value}'"
                    elif item.operator == 'contains':
                        item_condition = f"contains(item, '{item.value}')"
                    else:
                        item_condition = f"item {item.operator} '{item.value}'"
                    
                    part = f"{item.field}/{item.collection_function}(item: {item_condition})"
                else:
                    if isinstance(item.value, str):
                        part = f"{item.field} {item.operator} '{item.value}'"
                    else:
                        part = f"{item.field} {item.operator} {item.value}"
                
                parts.append(part)
                
            elif isinstance(item, FilterGroup):
                # Nested group
                nested_filter = self.build_complex_filter_tree(item)
                parts.append(f"({nested_filter})")
        
        return f" {filter_group.logical_operator} ".join(parts)
    
    def optimize_filter_expression(self, filter_expr: str) -> str:
        """
        Optimize a filter expression for better performance.
        
        Args:
            filter_expr: Original filter expression
            
        Returns:
            Optimized filter expression
        """
        # Remove unnecessary parentheses
        optimized = re.sub(r'\(\s*([^()]+)\s*\)', r'\1', filter_expr)
        
        # Simplify redundant conditions
        optimized = re.sub(r'(\w+\s+eq\s+\'[^\']+\')\s+and\s+\1', r'\1', optimized)
        
        # Move more selective conditions first (heuristic)
        # This is a simplified optimization - in practice, you'd analyze your data
        
        return optimized
    
    def validate_filter_syntax(self, filter_expr: str) -> Dict[str, Any]:
        """
        Validate filter expression syntax.
        
        Args:
            filter_expr: Filter expression to validate
            
        Returns:
            Validation results
        """
        issues = []
        warnings = []
        
        # Check for common syntax issues
        if filter_expr.count('(') != filter_expr.count(')'):
            issues.append("Mismatched parentheses")
        
        # Check for proper quoting
        if filter_expr.count("'") % 2 != 0:
            issues.append("Unmatched quotes")
        
        # Check for valid operators
        invalid_operators = re.findall(r'\s+(\w+)\s+', filter_expr)
        valid_ops = ['eq', 'ne', 'gt', 'ge', 'lt', 'le', 'and', 'or', 'not']
        for op in invalid_operators:
            if op not in valid_ops and not op.endswith('/any') and not op.endswith('/all'):
                warnings.append(f"Potentially invalid operator: {op}")
        
        return {
            'is_valid': len(issues) == 0,
            'issues': issues,
            'warnings': warnings,
            'complexity_score': self._calculate_complexity(filter_expr)
        }
    
    def _calculate_complexity(self, filter_expr: str) -> int:
        """Calculate complexity score for a filter expression."""
        score = 0
        score += filter_expr.count('and') * 1
        score += filter_expr.count('or') * 2
        score += filter_expr.count('/any') * 3
        score += filter_expr.count('/all') * 4
        score += filter_expr.count('(') * 1
        return score
    
    def search_with_complex_filter(self, filter_expr: str, search_text: str = "*",
                                 top: int = 10, include_facets: bool = False) -> Dict:
        """
        Execute a search with complex filter.
        
        Args:
            filter_expr: Complex filter expression
            search_text: Search query text
            top: Number of results to return
            include_facets: Whether to include facets
            
        Returns:
            Search results dictionary
        """
        try:
            # Validate filter first
            validation = self.validate_filter_syntax(filter_expr)
            if not validation['is_valid']:
                return {'error': f"Invalid filter: {validation['issues']}"}
            
            print(f"🔍 Executing complex search...")
            print(f"   Filter: {filter_expr}")
            print(f"   Complexity Score: {validation['complexity_score']}")
            
            start_time = time.time()
            
            search_params = {
                'search_text': search_text,
                'filter': filter_expr,
                'top': top,
                'include_total_count': True
            }
            
            if include_facets:
                search_params['facets'] = ['category', 'tags', 'rating']
            
            results = self.search_client.search(**search_params)
            
            execution_time = time.time() - start_time
            
            # Process results
            documents = list(results)
            total_count = results.get_count()
            facets = results.get_facets() if include_facets else None
            
            return {
                'documents': documents,
                'total_count': total_count,
                'execution_time': execution_time,
                'filter': filter_expr,
                'validation': validation,
                'facets': facets
            }
            
        except Exception as e:
            print(f"❌ Search error: {e}")
            return {'error': str(e)}
    
    def demonstrate_complex_filters(self):
        """Demonstrate various complex filtering scenarios."""
        print("🎯 Complex Filters Demonstration")
        print("=" * 50)
        
        examples = [
            {
                'name': 'Collection Any Filter',
                'filter': "tags/any(item: item eq 'premium') and rating ge 4.0",
                'description': 'Items with premium tag and rating >= 4.0'
            },
            {
                'name': 'Collection All Filter',
                'filter': "features/all(item: item eq 'available') and price le 100",
                'description': 'Items where all features are available and price <= 100'
            },
            {
                'name': 'Multi-Collection Filter',
                'filter': "tags/any(item: item eq 'new') and categories/any(item: item eq 'electronics')",
                'description': 'Items with new tag in electronics category'
            },
            {
                'name': 'Complex Nested Filter',
                'filter': "(tags/any(item: contains(item, 'special')) or rating gt 4.5) and (price ge 50 and price le 200)",
                'description': 'Special items or high-rated items in price range 50-200'
            },
            {
                'name': 'Geographic + Collection Filter',
                'filter': "geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 10 and amenities/any(item: item eq 'parking')",
                'description': 'Locations within 10km of Seattle with parking'
            }
        ]
        
        for example in examples:
            print(f"\n📋 {example['name']}")
            print(f"   Description: {example['description']}")
            print(f"   Filter: {example['filter']}")
            
            # Validate the filter
            validation = self.validate_filter_syntax(example['filter'])
            print(f"   Complexity Score: {validation['complexity_score']}")
            if validation['warnings']:
                print(f"   Warnings: {', '.join(validation['warnings'])}")
    
    def show_optimization_tips(self):
        """Display complex filter optimization tips."""
        print("\n🚀 Complex Filter Optimization Tips:")
        print("=" * 45)
        
        tips = [
            "Place most selective conditions first to reduce processing",
            "Use 'any()' instead of 'all()' when possible - it's more efficient",
            "Avoid deep nesting - flatten conditions when possible",
            "Combine range filters before collection filters",
            "Use field-specific indexes for frequently filtered collections",
            "Consider faceted navigation instead of complex OR conditions",
            "Cache complex filter results when appropriate",
            "Monitor query performance and adjust based on usage patterns",
            "Use search.score() to boost relevant results instead of complex filters",
            "Test filter selectivity with your actual data"
        ]
        
        for i, tip in enumerate(tips, 1):
            print(f"{i:2d}. {tip}")
    
    def analyze_filter_performance(self, filters: List[str]):
        """Analyze performance characteristics of different filters."""
        print("\n⚡ Filter Performance Analysis")
        print("=" * 35)
        
        print(f"{'Filter Type':<25} {'Complexity':<12} {'Performance':<15}")
        print("-" * 55)
        
        for filter_expr in filters:
            validation = self.validate_filter_syntax(filter_expr)
            complexity = validation['complexity_score']
            
            # Performance estimation based on complexity
            if complexity <= 5:
                performance = "Excellent"
            elif complexity <= 10:
                performance = "Good"
            elif complexity <= 20:
                performance = "Fair"
            else:
                performance = "Needs optimization"
            
            # Truncate long filters for display
            display_filter = filter_expr[:20] + "..." if len(filter_expr) > 20 else filter_expr
            
            print(f"{display_filter:<25} {complexity:<12} {performance:<15}")

def main():
    """Main demonstration function."""
    try:
        # Initialize the complex filter builder
        filter_builder = ComplexFilterBuilder()
        
        # Demonstrate complex filters
        filter_builder.demonstrate_complex_filters()
        
        # Show optimization tips
        filter_builder.show_optimization_tips()
        
        # Performance analysis with sample filters
        sample_filters = [
            "rating ge 4.0",
            "tags/any(item: item eq 'premium')",
            "tags/any(item: item eq 'new') and price le 100",
            "(tags/any(item: contains(item, 'special')) or rating gt 4.5) and (price ge 50 and price le 200)",
            "geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 10 and amenities/any(item: item eq 'parking') and features/all(item: item eq 'available')"
        ]
        
        filter_builder.analyze_filter_performance(sample_filters)
        
        # Example of building a complex filter programmatically
        print("\n🔧 Programmatic Filter Building Example:")
        print("=" * 45)
        
        # Create filter conditions
        conditions = [
            FilterCondition('tags', 'eq', 'premium', 'any'),
            FilterCondition('rating', 'ge', 4.0),
            FilterCondition('price', 'le', 200)
        ]
        
        complex_filter = filter_builder.build_nested_filter(conditions, 'and')
        print(f"Generated Filter: {complex_filter}")
        
        # Validate the generated filter
        validation = filter_builder.validate_filter_syntax(complex_filter)
        print(f"Validation: {'✅ Valid' if validation['is_valid'] else '❌ Invalid'}")
        print(f"Complexity Score: {validation['complexity_score']}")
        
        print("\n✅ Complex filters demonstration completed!")
        
    except Exception as e:
        print(f"❌ Error: {e}")

if __name__ == "__main__":
    main()

File: docs/beginner/module-06-filters-sorting/code-samples/javascript/07_complex_filters.js
/**
 * Complex Filters Example
 * 
 * This example demonstrates advanced filtering operations in Azure AI Search,
 * including collection filters, nested conditions, and complex logical expressions.
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');
require('dotenv').config();

class ComplexFiltersExample {
    constructor() {
        this.validateConfiguration();
        
        // Initialize search client
        const credential = new AzureKeyCredential(process.env.SEARCH_API_KEY);
        this.searchClient = new SearchClient(
            process.env.SEARCH_ENDPOINT,
            process.env.INDEX_NAME,
            credential
        );
    }

    validateConfiguration() {
        const requiredVars = ['SEARCH_ENDPOINT', 'SEARCH_API_KEY', 'INDEX_NAME'];
        const missingVars = requiredVars.filter(varName => !process.env[varName]);
        
        if (missingVars.length > 0) {
            throw new Error(`Missing required environment variables: ${missingVars.join(', ')}`);
        }
        
        console.log('✅ Configuration validated');
        console.log(`📍 Search Endpoint: ${process.env.SEARCH_ENDPOINT}`);
        console.log(`📊 Index Name: ${process.env.INDEX_NAME}`);
    }

    async demonstrateCollectionFilters() {
        console.log('\n📚 Collection Filters (any/all)');
        console.log('='.repeat(40));
        
        const collectionExamples = [
            {
                name: 'Items with any featured tag',
                filter: "tags/any(t: t eq 'featured')",
                description: 'Find items that have at least one "featured" tag'
            },
            {
                name: 'Items with wireless OR bluetooth tags',
                filter: "tags/any(t: t eq 'wireless' or t eq 'bluetooth')",
                description: 'Find items with connectivity-related tags'
            },
            {
                name: 'Items where all tags are non-empty',
                filter: "tags/all(t: t ne null and t ne '')",
                description: 'Find items where every tag has a meaningful value'
            },
            {
                name: 'Items with premium-related tags',
                filter: "tags/any(t: contains(t, 'premium') or contains(t, 'luxury') or contains(t, 'pro'))",
                description: 'Find items with any premium-related tags'
            }
        ];

        for (const example of collectionExamples) {
            console.log(`\n📋 ${example.name}`);
            console.log(`   Description: ${example.description}`);
            console.log(`   Filter: ${example.filter}`);
            
            try {
                const searchResults = await this.searchClient.search('*', {
                    filter: example.filter,
                    top: 3,
                    select: ['id', 'name', 'tags', 'category', 'price']
                });

                const results = [];
                for await (const result of searchResults.results) {
                    results.push(result.document);
                }

                console.log(`   Results: ${results.length} items found`);
                
                results.slice(0, 2).forEach((result, index) => {
                    const name = result.name || 'N/A';
                    const category = result.category || 'N/A';
                    const price = result.price || 'N/A';
                    const tags = result.tags ? result.tags.join(', ') : 'No tags';
                    console.log(`     ${index + 1}. ${name} (${category}) - $${price}`);
                    console.log(`        Tags: ${tags}`);
                });
                
            } catch (error) {
                console.log(`   ❌ Error: ${error.message}`);
            }
        }
    }

    async demonstrateNestedComplexFilters() {
        console.log('\n🔗 Nested Complex Filters');
        console.log('='.repeat(40));
        
        const nestedExamples = [
            {
                name: 'Complex product filtering',
                filter: "(category eq 'Electronics' and price gt 100) or (category eq 'Books' and rating ge 4.5)",
                description: 'Find expensive electronics OR highly-rated books'
            },
            {
                name: 'Multi-criteria premium items',
                filter: "(brand eq 'Apple' or brand eq 'Samsung') and (price gt 500 and rating gt 4.0) and tags/any(t: t eq 'premium')",
                description: 'Find premium items from top brands with high price and rating'
            },
            {
                name: 'Availability and quality filter',
                filter: "inStock eq true and ((rating ge 4.0 and price le 200) or (rating ge 4.5 and price le 500))",
                description: 'Find in-stock items that are either good value or high quality'
            },
            {
                name: 'Geographic and category filter',
                filter: "(category eq 'Restaurant' or category eq 'Hotel') and geo.distance(location, geography'POINT(-122.335167 47.608013)') le 25",
                description: 'Find restaurants or hotels within 25km of Seattle'
            }
        ];

        for (const example of nestedExamples) {
            console.log(`\n📋 ${example.name}`);
            console.log(`   Description: ${example.description}`);
            console.log(`   Filter: ${example.filter}`);
            
            try {
                const searchResults = await this.searchClient.search('*', {
                    filter: example.filter,
                    top: 3,
                    select: ['id', 'name', 'category', 'brand', 'price', 'rating', 'inStock', 'tags', 'location']
                });

                const results = [];
                for await (const result of searchResults.results) {
                    results.push(result.document);
                }

                console.log(`   Results: ${results.length} items found`);
                
                results.slice(0, 2).forEach((result, index) => {
                    const name = result.name || 'N/A';
                    const category = result.category || 'N/A';
                    const brand = result.brand || 'N/A';
                    const price = result.price || 'N/A';
                    const rating = result.rating || 'N/A';
                    const inStock = result.inStock !== undefined ? result.inStock : 'N/A';
                    console.log(`     ${index + 1}. ${name} (${category}) - ${brand} - $${price} - ${rating}⭐ - Stock: ${inStock}`);
                });
                
            } catch (error) {
                console.log(`   ❌ Error: ${error.message}`);
            }
        }
    }

    async demonstrateSearchWithComplexFilters() {
        console.log('\n🔍 Search Queries with Complex Filters');
        console.log('='.repeat(40));
        
        const searchFilterExamples = [
            {
                name: 'Search "laptop" with complex filters',
                searchText: 'laptop computer',
                filter: "(category eq 'Electronics' and price ge 500 and price le 2000) and (rating ge 4.0 or tags/any(t: t eq 'bestseller'))",
                description: 'Search for laptops with price range and quality filters'
            },
            {
                name: 'Search "wireless" with availability',
                searchText: 'wireless bluetooth',
                filter: "inStock eq true and (tags/any(t: t eq 'wireless' or t eq 'bluetooth')) and price le 300",
                description: 'Search wireless products that are in stock and affordable'
            },
            {
                name: 'Search "premium" with brand filter',
                searchText: 'premium quality',
                filter: "(brand eq 'Apple' or brand eq 'Samsung' or brand eq 'Sony') and rating ge 4.5",
                description: 'Search premium products from trusted brands'
            }
        ];

        for (const example of searchFilterExamples) {
            console.log(`\n📋 ${example.name}`);
            console.log(`   Description: ${example.description}`);
            console.log(`   Search Text: "${example.searchText}"`);
            console.log(`   Filter: ${example.filter}`);
            
            try {
                const searchResults = await this.searchClient.search(example.searchText, {
                    filter: example.filter,
                    top: 3,
                    select: ['id', 'name', 'category', 'brand', 'price', 'rating', 'tags', 'inStock'],
                    orderBy: ['search.score() desc', 'rating desc']
                });

                const results = [];
                for await (const result of searchResults.results) {
                    results.push(result.document);
                }

                console.log(`   Results: ${results.length} items found`);
                
                results.slice(0, 2).forEach((result, index) => {
                    const name = result.name || 'N/A';
                    const category = result.category || 'N/A';
                    const brand = result.brand || 'N/A';
                    const price = result.price || 'N/A';
                    const rating = result.rating || 'N/A';
                    console.log(`     ${index + 1}. ${name} (${category}) - ${brand} - $${price} - ${rating}⭐`);
                });
                
            } catch (error) {
                console.log(`   ❌ Error: ${error.message}`);
            }
        }
    }

    async demonstrateAdvancedCollectionScenarios() {
        console.log('\n🎯 Advanced Collection Scenarios');
        console.log('='.repeat(40));
        
        const advancedCollectionExamples = [
            {
                name: 'Items with multiple specific tags',
                filter: "tags/any(t: t eq 'wireless') and tags/any(t: t eq 'premium')",
                description: 'Find items that have both wireless AND premium tags'
            },
            {
                name: 'Items missing essential tags',
                filter: "not tags/any(t: t eq 'featured' or t eq 'bestseller' or t eq 'premium')",
                description: 'Find items without any promotional tags'
            },
            {
                name: 'Items with tag count implications',
                filter: "tags/any() and not tags/any(t: t eq null or t eq '')",
                description: 'Find items that have tags and all tags are meaningful'
            },
            {
                name: 'Complex tag pattern matching',
                filter: "tags/any(t: startswith(t, 'color-') or startswith(t, 'size-') or startswith(t, 'material-'))",
                description: 'Find items with attribute-based tags'
            }
        ];

        for (const example of advancedCollectionExamples) {
            console.log(`\n📋 ${example.name}`);
            console.log(`   Description: ${example.description}`);
            console.log(`   Filter: ${example.filter}`);
            
            try {
                const searchResults = await this.searchClient.search('*', {
                    filter: example.filter,
                    top: 3,
                    select: ['id', 'name', 'tags', 'category', 'price']
                });

                const results = [];
                for await (const result of searchResults.results) {
                    results.push(result.document);
                }

                console.log(`   Results: ${results.length} items found`);
                
                results.slice(0, 2).forEach((result, index) => {
                    const name = result.name || 'N/A';
                    const category = result.category || 'N/A';
                    const price = result.price || 'N/A';
                    const tags = result.tags ? result.tags.join(', ') : 'No tags';
                    console.log(`     ${index + 1}. ${name} (${category}) - $${price}`);
                    console.log(`        Tags: ${tags}`);
                });
                
            } catch (error) {
                console.log(`   ❌ Error: ${error.message}`);
            }
        }
    }

    buildComplexFilter(criteria) {
        /**
         * Build a complex filter from structured criteria
         * @param {Object} criteria - Filter criteria object
         * @returns {string} Complex filter expression
         */
        const filters = [];
        
        // Category filters
        if (criteria.categories && criteria.categories.length > 0) {
            const categoryFilters = criteria.categories.map(cat => `category eq '${cat}'`);
            filters.push(`(${categoryFilters.join(' or ')})`);
        }
        
        // Price range
        if (criteria.minPrice !== undefined || criteria.maxPrice !== undefined) {
            const priceFilters = [];
            if (criteria.minPrice !== undefined) {
                priceFilters.push(`price ge ${criteria.minPrice}`);
            }
            if (criteria.maxPrice !== undefined) {
                priceFilters.push(`price le ${criteria.maxPrice}`);
            }
            filters.push(`(${priceFilters.join(' and ')})`);
        }
        
        // Rating filter
        if (criteria.minRating !== undefined) {
            filters.push(`rating ge ${criteria.minRating}`);
        }
        
        // Brand filters
        if (criteria.brands && criteria.brands.length > 0) {
            const brandFilters = criteria.brands.map(brand => `brand eq '${brand}'`);
            filters.push(`(${brandFilters.join(' or ')})`);
        }
        
        // Tag filters
        if (criteria.requiredTags && criteria.requiredTags.length > 0) {
            const tagFilters = criteria.requiredTags.map(tag => `tags/any(t: t eq '${tag}')`);
            filters.push(`(${tagFilters.join(' and ')})`);
        }
        
        if (criteria.anyTags && criteria.anyTags.length > 0) {
            const anyTagFilters = criteria.anyTags.map(tag => `t eq '${tag}'`);
            filters.push(`tags/any(t: ${anyTagFilters.join(' or ')})`);
        }
        
        // Availability
        if (criteria.inStock !== undefined) {
            filters.push(`inStock eq ${criteria.inStock}`);
        }
        
        // Geographic filter
        if (criteria.location) {
            const { lat, lon, radiusKm } = criteria.location;
            const point = `geography'POINT(${lon} ${lat})'`;
            filters.push(`geo.distance(location, ${point}) le ${radiusKm}`);
        }
        
        // Date filter
        if (criteria.modifiedAfter) {
            const date = criteria.modifiedAfter instanceof Date ? 
                criteria.modifiedAfter.toISOString() : criteria.modifiedAfter;
            filters.push(`lastModified ge ${date}`);
        }
        
        return filters.length > 0 ? filters.join(' and ') : null;
    }

    async demonstrateDynamicComplexFiltering() {
        console.log('\n🏗️ Dynamic Complex Filter Building');
        console.log('='.repeat(40));
        
        const complexScenarios = [
            {
                name: 'Premium electronics search',
                criteria: {
                    categories: ['Electronics', 'Computers'],
                    minPrice: 500,
                    maxPrice: 2000,
                    minRating: 4.0,
                    brands: ['Apple', 'Samsung', 'Sony'],
                    anyTags: ['premium', 'bestseller', 'featured'],
                    inStock: true
                }
            },
            {
                name: 'Local restaurant search',
                criteria: {
                    categories: ['Restaurant', 'Food'],
                    minRating: 4.0,
                    location: { lat: 47.608013, lon: -122.335167, radiusKm: 10 },
                    anyTags: ['delivery', 'takeout', 'dine-in']
                }
            },
            {
                name: 'Recent budget items',
                criteria: {
                    maxPrice: 100,
                    minRating: 3.5,
                    modifiedAfter: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000),
                    inStock: true,
                    anyTags: ['sale', 'discount', 'budget']
                }
            }
        ];

        for (const scenario of complexScenarios) {
            console.log(`\n📋 ${scenario.name}`);
            const filter = this.buildComplexFilter(scenario.criteria);
            
            console.log(`   Generated Filter: ${filter}`);
            console.log(`   Criteria: ${JSON.stringify(scenario.criteria, null, 2)}`);
            
            if (filter) {
                try {
                    const searchResults = await this.searchClient.search('*', {
                        filter: filter,
                        top: 2,
                        select: ['id', 'name', 'category', 'brand', 'price', 'rating', 'tags', 'inStock', 'location']
                    });

                    const results = [];
                    for await (const result of searchResults.results) {
                        results.push(result.document);
                    }

                    console.log(`   Results: ${results.length} items found`);
                    
                    results.forEach((result, index) => {
                        const name = result.name || 'N/A';
                        const category = result.category || 'N/A';
                        const brand = result.brand || 'N/A';
                        const price = result.price || 'N/A';
                        const rating = result.rating || 'N/A';
                        console.log(`     ${index + 1}. ${name} (${category}) - ${brand} - $${price} - ${rating}⭐`);
                    });
                    
                } catch (error) {
                    console.log(`   ❌ Error: ${error.message}`);
                }
            }
        }
    }

    demonstrateComplexFilterBestPractices() {
        console.log('\n💡 Complex Filter Best Practices');
        console.log('='.repeat(40));
        
        console.log('\n1. Logical Operator Precedence');
        console.log('   ✅ Use parentheses to group conditions clearly');
        console.log('   ✅ Remember: NOT > Comparison > AND > OR');
        console.log('   ✅ Example: (A and B) or (C and D)');
        console.log('   ❌ Avoid: A and B or C and D (ambiguous)');
        
        console.log('\n2. Collection Filter Rules');
        console.log('   ✅ Use any() for "at least one" conditions');
        console.log('   ✅ Use all() for "every item" conditions');
        console.log('   ✅ Combine multiple any() for AND logic between collections');
        console.log('   ❌ Don\'t nest collection operators');
        
        console.log('\n3. Performance Optimization');
        console.log('   ✅ Put most selective filters first');
        console.log('   ✅ Use simple filters before complex ones');
        console.log('   ✅ Limit the number of OR conditions');
        console.log('   ❌ Avoid deeply nested expressions');
        
        console.log('\n4. Filter Complexity Management');
        console.log('   ✅ Break complex filters into smaller parts');
        console.log('   ✅ Use helper functions to build filters');
        console.log('   ✅ Test individual filter components');
        console.log('   ✅ Document complex filter logic');
        
        console.log('\n5. Common Complex Patterns');
        console.log('   🔍 Multi-category: (cat eq \'A\' or cat eq \'B\') and price gt 100');
        console.log('   🏷️ Tag combinations: tags/any(t: t eq \'X\') and tags/any(t: t eq \'Y\')');
        console.log('   📍 Geo + filters: geo.distance(...) le 10 and category eq \'Restaurant\'');
        console.log('   🔗 Search + filter: Full-text search with complex filter criteria');
        
        console.log('\n6. Error Prevention');
        console.log('   ✅ Validate all input values before building filters');
        console.log('   ✅ Escape special characters in string values');
        console.log('   ✅ Handle null and empty values appropriately');
        console.log('   ✅ Test with edge cases and boundary conditions');
        
        console.log('\n7. Debugging Complex Filters');
        console.log('   ✅ Test individual filter components separately');
        console.log('   ✅ Use simple test data to verify logic');
        console.log('   ✅ Check operator precedence with parentheses');
        console.log('   ✅ Validate OData syntax with simple tools');
    }

    async run() {
        console.log('🚀 Complex Filters Example');
        console.log('='.repeat(50));
        
        try {
            await this.demonstrateCollectionFilters();
            await this.demonstrateNestedComplexFilters();
            await this.demonstrateSearchWithComplexFilters();
            await this.demonstrateAdvancedCollectionScenarios();
            await this.demonstrateDynamicComplexFiltering();
            this.demonstrateComplexFilterBestPractices();
            
            console.log('\n✅ Complex filters example completed successfully!');
            console.log('\nKey takeaways:');
            console.log('- Use collection filters (any/all) for array fields');
            console.log('- Group complex conditions with parentheses for clarity');
            console.log('- Combine search queries with complex filters for precision');
            console.log('- Build dynamic filters from structured criteria objects');
            console.log('- Optimize performance by ordering filters by selectivity');
            console.log('- Test complex filters thoroughly with various data scenarios');
            
        } catch (error) {
            console.log(`\n❌ Example failed: ${error.message}`);
            throw error;
        }
    }
}

async function main() {
    const example = new ComplexFiltersExample();
    try {
        await example.run();
    } catch (error) {
        console.error(`Application failed: ${error.message}`);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = ComplexFiltersExample;